File: leetcode/53/c++/solution.cpp
/*
    53. Maximum Subarray
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    int maxSubArray(vector<int>& nums) {
        int max_sum = nums[0], sum = 0;

        for (size_t i = 0; i < nums.size(); ++i) {
            sum += nums[i];

            if (max_sum < sum) {
                max_sum = sum;
            }

            if (sum < 0) {
                sum = 0;
            }
        }

        return max_sum;
    }
};

File: leetcode/53/rust/solution.rs
/*
    53. Maximum Subarray
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut max_sum: i32 = nums[0];
        let mut local_sum: i32 = 0;
        for num_i in nums.iter() {
            local_sum += *num_i;
            max_sum = max_sum.max(local_sum);
            local_sum = local_sum.max(0);
        }
        max_sum
    }
}